go/src/koding/kites/kloud/provider/softlayer/stack.go

```diff
@@ -0,0 +1,213 @@
+package softlayer
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"strconv"
+	"text/template"
+
+	"koding/kites/kloud/stack"
+	"koding/kites/kloud/stack/provider"
+	"koding/kites/kloud/userdata"
+
+	softlayerGo "github.com/maximilien/softlayer-go/client"
+)
+
+//go:generate $GOPATH/bin/go-bindata -mode 420 -modtime 1475345133 -pkg softlayer -o bootstrap.json.tmpl.go bootstrap.json.tmpl
+//go:generate go fmt bootstrap.json.tmpl.go
+
+var (
+	_ provider.Stack = (*Stack)(nil)
+	_ stack.Stacker  = (*Stack)(nil)
+)
+
+func mustAsset(s string) string {
+	a, err := Asset(s)
+	if err != nil {
+		panic(err)
+	}
+	return string(a)
+}
+
+var BootstrapTemplate = template.Must(template.New("").Parse(mustAsset("bootstrap.json.tmpl")))
+
+type bootstrapVars struct {
+	KeyName   string
+	PublicKey string
+}
+
+// Responsible for building/updating/destroying Softlayer's stack
+type Stack struct {
+	*provider.BaseStack
+}
+
+// Verifies the given Softlayer credential.
+// Compared against what is defined in schema.Credential.
+func (s *Stack) VerifyCredential(c *stack.Credential) error {
+	cred, ok := c.Credential.(*Credential)
+	if !ok {
+		return fmt.Errorf("credential is not of type softlayer.Credential: %T", c.Credential)
+	}
+
+	if err := cred.Valid(); err != nil {
+		return err
+	}
+
+	// Do a quick check to verify the credentials are valid for
+	// communitcating with the SoftLayer apis
+	client := softlayerGo.NewSoftLayerClient(cred.Username, cred.ApiKey)
+	service, err := client.GetSoftLayer_Virtual_Guest_Service()
+	if err != nil {
+		return err
+	}
+	_, err = service.GetObject(1234567)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *Stack) BootstrapArg() *stack.BootstrapRequest {
+	return s.BaseStack.Arg.(*stack.BootstrapRequest)
+}
+
+// BootstrapTemplates creates the template that is used to generate default
+// output values, to bootstrap a SoftLayer stack
+func (s *Stack) BootstrapTemplates(c *stack.Credential) ([]*stack.Template, error) {
+	vs := &bootstrapVars{
+		KeyName: fmt.Sprintf(
+			"koding-%s",
+			c.Identifier,
+		),
+		PublicKey: s.Keys.PublicKey,
+	}
+
+	var bootstrap bytes.Buffer
+	if err := BootstrapTemplate.Execute(&bootstrap, vs); err != nil {
+		return nil, err
+	}
+
+	// NOTE: Should we shadow out private vars here?
```

Comment:
Not needed. Shadowing is there to prevent leaking delagated credentials.

If you have an admin and a user within your team, and user should now know Softlayer credentials used to create resources, then admin creates a credential and shares it with the user. Since the template is augemented server-side, user never sees the real credentials used.

The shadowing is used to prevent malicious user to extract credentials from template, e.g. by echoing them to a file during vm creating.

Since user has no access currently to bootstrap templates, no need for shadowing.